bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/row_range.h"
+
+#include <iostream>
+#include <thread>
+
+#include <absl/memory/memory.h>
+#include <absl/strings/str_cat.h>
+
+namespace btproto = ::google::bigtable::v2;
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+const std::string RowRange::SAMPLE_KEY = "";
+;
+
+/**
+ * The contructor for RowRange: [begin, end).
+ * In the constructor and all the builder methods, We don't check
+ * if begin <= end. This anomaly is caught during range manipulation
+ * (i.e., intersect, contains)
+ */
+RowRange::RowRange(std::string begin, std::string end) {
+  range_.set_start_key_closed(std::move(begin));
+  if (!end.empty()) range_.set_end_key_open(std::move(end));
+}
+
+/**
+ * An empty range satisfies the following conditions:
+ *   - Any semi-open or open range [x, x), (x,x], (x,x) where x could be empty.
+ *   - 'end' is set (otherwise it would be an infinite range).
+ *   - Not a closed interval [x,x] with non-empty x.
+ *
+ * We use the right-open [x,x) as the default empty range.
+ * For convenience and the readability of unit tests,
+ * we set the SAMPLE_KEY = "".
+ * A set end_key (even to "") makes a finite range.
+ * (Only this method is allowed set end_key to "". A user can't create a
+ * finite range with end_key "", which is reserved for (semi-)infinite range).
+ */
+RowRange RowRange::empty() {
+  RowRange tmp;
+  tmp.range_.set_start_key_closed(RowRange::SAMPLE_KEY);
+  // Only this method can set end_key as "".
+  tmp.range_.set_end_key_open(RowRange::SAMPLE_KEY);
+  return tmp;
+}
+
+/**
+ * Infinite range has the following properties:
+ *   - 'end' value is not set.
+ *   - 'start' could be  don't cared.
+ *      The proto also has the property that, an unset start key will have
+ *      value "", inclusive.
+ *  Gives the range ["..NOT_SET).
+ *
+ *  Note: We support right-sided (or semi) inifinity, where start key can be set
+ * to any value: [x..NOT_SET); an empty end key implies semi-infinite range.
+ */
+RowRange RowRange::infinite_range() { return RowRange(); }
+
+/**
+ * Default one, right-open: [begin, end).
```

Comment:
If this is intended as a Doxygen comment, remember that the audience is users of this class/function.  It should say what the function does, e.g.,

```
/**
  * Return a RowRange representing the range [@p begin, @p end).
  *
  * Note that the @p end point is not included in the range.  This function makes no attempt
  * at validating the range locally.  Invalid or empty ranges might be rejected by the server.
  */
```

Reply:
Also, this comment should be in the header file.

Reply:
Thanks. I think I need some  massaging on the comments. Let me shift the comments to header.  
Once we place the comments in header,  should the .cc file have no function-level comments ? I will look up in the cs/ code base for samples.